src/main/java/spireMapOverhaul/zones/smithsFolly/SmithsFolly.java

```diff
@@ -0,0 +1,179 @@
+                    }
+                    break;
+                }
+                case 3: {
+                    for (final AbstractMonster m : abstractRoom.monsters.monsters) {
+                        AbstractDungeon.actionManager.addToBottom(new ApplyPowerAction(m, m, new RegenerateMonsterPower(m, 1 + AbstractDungeon.actNum * 2), 1 + AbstractDungeon.actNum * 2));
+                    }
+                    break;
+                }
+            }
+        }
+    }
+
+    @Override
+    public RewardItem getAdditionalReward() {
+        if (!AbstractDungeon.getCurrMapNode().hasEmeraldKey && AbstractDungeon.getCurrRoom() instanceof MonsterRoomElite) {
+            if (!Settings.hasEmeraldKey) {
+                return new RewardItem(null, RewardItem.RewardType.EMERALD_KEY);
+            }
+        }
+        return null;
+    }
+
+    private void upgradeRarestCard() {
+        AbstractPlayer player = AbstractDungeon.player;
+        if (player == null) return;
+
+        // Define the rarity order
+        List<AbstractCard.CardRarity> rarityOrder = Arrays.asList(
+                AbstractCard.CardRarity.SPECIAL,
+                AbstractCard.CardRarity.RARE,
+                AbstractCard.CardRarity.UNCOMMON,
+                AbstractCard.CardRarity.COMMON,
+                AbstractCard.CardRarity.BASIC
+        );
+
+        // Find the upgradable cards, grouped by rarity
+        Map<AbstractCard.CardRarity, List<AbstractCard>> upgradableCardsByRarity = player.masterDeck.group.stream()
+                .filter(AbstractCard::canUpgrade)
+                .collect(Collectors.groupingBy(card -> card.rarity));
+
+        // Find the rarest category with upgradable cards
+        for (AbstractCard.CardRarity rarity : rarityOrder) {
+            List<AbstractCard> upgradableCards = upgradableCardsByRarity.getOrDefault(rarity, Collections.emptyList());
+            if (!upgradableCards.isEmpty()) {
+                // Shuffle and upgrade a random card from the rarest category
+                Collections.shuffle(upgradableCards, new Random(AbstractDungeon.miscRng.randomLong()));
+                AbstractCard cardToUpgrade = upgradableCards.get(0);
+                cardToUpgrade.upgrade();
+                AbstractDungeon.player.bottledCardUpgradeCheck(cardToUpgrade);
+
+                // Show upgrade effect (optional)
+                AbstractDungeon.topLevelEffects.add(new ShowCardBrieflyEffect(cardToUpgrade.makeStatEquivalentCopy()));
+                AbstractDungeon.topLevelEffects.add(new UpgradeShineEffect((float) Settings.WIDTH / 2.0F, (float)Settings.HEIGHT / 2.0F));
+                break;
+            }
+        }
+    }
+
+
+    @Override
+    protected boolean canIncludeEarlyRows() {
+        return false;
+    }
+
+    @Override
+    public float zoneSpecificEventRate() {
```

Comment:
Could also add the cursed forge (or whatever it's name is) from base game if you wanted to.
```
@Override
    public Set<String> addSpecificEvents() {
        Set<String> baseGameEvents = new HashSet<>();
        baseGameEvents.add(TheEvent.ID);
        return baseGameEvents;
    }
```